libs/schemas/src/Data/Schema.hs

```diff
@@ -0,0 +1,559 @@
+
+instance Monoid doc => Monoid (SchemaP doc v v' a b) where
+  mempty = SchemaP mempty mempty mempty
+
+instance Profunctor (SchemaP doc v v') where
+  dimap f g (SchemaP d i o) =
+    SchemaP (dimap f g d) (dimap f g i) (dimap f g o)
+
+instance Choice (SchemaP doc v v') where
+  left' (SchemaP d i o) = SchemaP (left' d) (left' i) (left' o)
+  right' (SchemaP d i o) = SchemaP (right' d) (right' i) (right' o)
+
+instance HasDoc (SchemaP doc v v' a b) (SchemaP doc' v v' a b) doc doc' where
+  doc = lens schemaDoc $ \(SchemaP d i o) d' -> SchemaP (set doc d' d) i o
+
+withParser :: SchemaP doc v w a b -> (b -> A.Parser b') -> SchemaP doc v w a b'
+withParser (SchemaP (SchemaDoc d) (SchemaIn p) (SchemaOut o)) q =
+  SchemaP (SchemaDoc d) (SchemaIn (p >=> q)) (SchemaOut o)
+
+type SchemaP' doc v v' a = SchemaP doc v v' a a
+
+type ObjectSchema doc a = SchemaP' doc A.Object [A.Pair] a
+
+type ValueSchema doc a = SchemaP' doc A.Value A.Value a
+
+type EnumSchema doc a = SchemaP' doc Text (Alt Maybe Text) a
+
+schemaDoc :: SchemaP ss v m a b -> ss
+schemaDoc (SchemaP (SchemaDoc d) _ _) = d
+
+schemaIn :: SchemaP doc v v' a b -> v -> A.Parser b
+schemaIn (SchemaP _ (SchemaIn i) _) = i
+
+schemaOut :: SchemaP ss v m a b -> a -> Maybe m
+schemaOut (SchemaP _ _ (SchemaOut o)) = o
+
+-- | A schema for a one-field JSON object.
+field :: HasField doc' doc => Text -> ValueSchema doc' a -> ObjectSchema doc a
+field name sch = SchemaP (SchemaDoc s) (SchemaIn r) (SchemaOut w)
+  where
+    r obj = A.explicitParseField (schemaIn sch) obj name
+    w x = do
+      v <- schemaOut sch x
+      pure [name A..= v]
+
+    s = mkField name (schemaDoc sch)
+
+-- | Like 'field', but apply an arbitrary function to the
+-- documentation of the field.
+field' ::
+  HasField doc' doc =>
+  Text ->
+  (doc' -> doc') ->
+  ValueSchema doc' a ->
+  ObjectSchema doc a
+field' name modify sch = field name (over doc modify sch)
+
+-- | Change the input type of a schema.
+(.=) :: Profunctor p => (a -> a') -> p a' b -> p a b
+(.=) = lmap
+
+-- | Change the input and output types of a schema via a prism.
+tag :: Prism b b' a a' -> SchemaP ss v m a a' -> SchemaP ss v m b b'
+tag f = rmap runIdentity . f . rmap Identity
+
+-- | A schema for a JSON object.
+--
+-- This can be used to convert a combination of schemas obtained using
+-- 'field' into a single schema for a JSON object.
+object :: HasObject doc doc' => Text -> ObjectSchema doc a -> ValueSchema doc' a
+object name sch = SchemaP (SchemaDoc s) (SchemaIn r) (SchemaOut w)
+  where
+    r = A.withObject (T.unpack name) (schemaIn sch)
+    w x = A.object <$> schemaOut sch x
+    s = mkObject name (schemaDoc sch)
+
+-- | Like 'object', but apply an arbitrary function to the
+-- documentation of the resulting object.
+object' ::
+  HasObject doc doc' =>
+  Text ->
+  (doc' -> doc') ->
+  ObjectSchema doc a ->
+  ValueSchema doc' a
+object' name modify sch = over doc modify (object name sch)
```

Comment:
This prime makes this function harder to discover. But I can also not come up with nicer name, here are my attempts:
- `ObjectWithModifiedDoc`
- `overDocObject`

Another thing we can do is just expect users to use the lens, or does it make things more complicated?

Reply:
Using the lens directly is possible, but slightly akwkard, because the super-general lens types make type inference fail if you write it in the obvious way. The reason is that the `doc` lens can change the underlying `doc` type, whereas within `object'` the modifier function is constrained to be an endomorphism `doc' -> doc'`.

Besides renaming, another possible solution would be to make the general `doc` lens internal, rename it, and export a type-restricted `doc` lens. That hopefully should make type inference work in most cases.

If that doesn't work, I'll rename `object'` to one of your suggestions.

Reply:
I ended up renaming it to `objectWithDocModifier`. It's useful to have it as a convenience function instead of messing with lenses and having to add a bunch of extra parentheses. Same for `field'`.

Reply:
Sounds good. :+1: